annotation_pipeline/segment.py

```diff
@@ -1,101 +1,149 @@
-            sp_inds_list.append(np.ones_like(mzs_) * sp_idx)
-            mzs_list.append(mzs_)
-            ints_list.append(ints_)
-            estimated_size_mb += 2 * mzs_.nbytes + ints_.nbytes
-            curr_sp_i += 1
-            if estimated_size_mb > max_size:
-                yield sp_inds_list, mzs_list, ints_list
-                sp_inds_list, mzs_list, ints_list = [], [], []
+        # Iterate in the same order that intensities are laid out in the file, hopefully this will
+        # prevent fragmented read patterns
+        for sp_i in np.argsort(imzml_reader.intensityOffsets):
+            spectrum_size = imzml_reader.mzLengths[sp_i] * row_size
+            if estimated_size_mb + spectrum_size > MAX_CHUNK_SIZE:
                 estimated_size_mb = 0
+                yield np.array(chunk_sp_inds)
+                chunk_sp_inds = []
+
+            estimated_size_mb += spectrum_size
+            chunk_sp_inds.append(sp_i)
+
+        if chunk_sp_inds:
+            yield np.array(chunk_sp_inds)
+
+    def upload_chunk(ibm_cos, ch_i):
+        chunk_sp_inds = chunks[ch_i]
+        # Get imzml_reader from COS because it's too big to include via pywren captured vars
+        imzml_reader = pickle.loads(read_object_with_retry(ibm_cos, config["storage"]["ds_bucket"],
+                                              input_data["ds_imzml_reader"]))
+        n_spectra = sum(imzml_reader.mzLengths[sp_i] for sp_i in chunk_sp_inds)
+        sp_mz_int_buf = np.zeros((n_spectra, 3), dtype=imzml_reader.mzPrecision)
+
+        chunk_start = 0
+        for sp_i, mzs, ints in get_spectra(input_data['ibd_path'], imzml_reader, chunk_sp_inds):
+            chunk_end = chunk_start + len(mzs)
+            sp_mz_int_buf[chunk_start:chunk_end, 0] = sp_id_to_idx[sp_i]
+            sp_mz_int_buf[chunk_start:chunk_end, 1] = mzs
+            sp_mz_int_buf[chunk_start:chunk_end, 2] = ints
+            chunk_start = chunk_end
+
+        by_mz = np.argsort(sp_mz_int_buf[:, 1])
+        sp_mz_int_buf = sp_mz_int_buf[by_mz]
+        del by_mz
 
-        if len(sp_inds_list) > 0:
-            yield sp_inds_list, mzs_list, ints_list
-
-    def _upload_chunk(ch_i, sp_mz_int_buf):
         chunk = msgpack.dumps(sp_mz_int_buf)
         key = f'{input_data["ds_chunks"]}/{ch_i}.msgpack'
         size = sys.getsizeof(chunk) * (1 / 1024 ** 2)
         logger.info(f'Uploading spectra chunk {ch_i} - %.2f MB' % size)
-        cos_client.put_object(Bucket=config["storage"]["ds_bucket"],
-                              Key=key,
-                              Body=chunk)
+        ibm_cos.put_object(Bucket=config["storage"]["ds_bucket"],
+                           Key=key,
+                           Body=chunk)
         logger.info(f'Spectra chunk {ch_i} finished')
         return key
 
-    max_size = 512 * 1024 ** 2  # 512MB
-    chunk_it = chunk_size(coordinates, max_size)
+    chunks = list(plan_chunks())
+    memory_capacity_mb = 3072
```

Comment:
Unfortunately I couldn't get memory usage lower than 2.3GB with huge4. In theory it should be possible to for this function to run with only `2 * MAX_CHUNK_SIZE`, but it seems like the memory allocated by `requests.get` in `read_ranges_from_url` can't be re-used for numpy arrays.

Reply:
I'm not sure if it will fit well in this case, but `ibmboto3` interface allows using ranges as well, maybe it does it in an optimized way that will reduce memory usage.
when using `cos_client.get_object` you can specify byte ranges by:
```python
cos_client.get_object(Bucket='...', Key='...', Range='bytes 0-499')
```
more details about ranges format can be found here: https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.35

Reply:
@omerb01 I didn't realize that `ibmboto3` could be used for accessing objects in other accounts' public buckets, but I've just tested it and it seems to work. I'll update the code to use `ibmboto3` instead of HTTP.

Reply:
Actually, there's a problem with directly accessing it through `ibmboto3` - regions. 

These notebooks are public demonstrations of PyWren, and because of that, we want to keep them easily runnable by anybody with the data provided through my public bucket. Accessing this data by HTTP only requires one parameter - the URL, but accessing it with `ibmboto3` requires a `cos_client` that was configured with the correct `endpoint` for the region, plus a `Bucket` and `Key`. 

It definitely is possible to support this, but I think it's too much complexity to pass 3 different parameters and somehow reconfigure the `cos_client` instance to always use the `eu-de` region for this bucket...

I'm going to push a commit that includes my changes to use `ibmboto3` in case you want to experiment, but I'm also going to revert these changes because I don't think they can be used in their current state, because they fail without much explanation if your main COS configuration is set to a different region.

Reply:
@LachlanStuart thanks, just for clarity, does this approach solve the memory issue that you observed? if it does, please commit this function as well and let me investigate it a bit after the holidays.

Reply:
@omerb01 the `ibmboto3` version of the function is in this commit: https://github.com/metaspace2020/pywren-annotation-pipeline/pull/63/commits/23adb53a114a083e8c61153565319297086c229e

I'm not actually sure if it solves the memory issue. PyWren stopped reporting peak memory usage after I updated, so I'll try to figure out what happened and make a bug report.

Reply:
@LachlanStuart I fixed 'memory peak measurement' in https://github.com/pywren/pywren-ibm-cloud/pull/288

Reply:
Thanks @JosepSampe .